keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,538 @@
+This field is introduced within ServicePort rather than ServiceSpec, because a ServicePort entry is needed to infer the protocol.
+
+This field will be supported for Service Type=ClusterIP and ServiceType=LoadBalancer. This is not applicable to ExternalName services. NodePort services are not supported.
+A NodePort service accepts traffic on a given port on the Node, redirecting it to the specified targetPort of the service endpoints. Support AllPorts for a NodePort service means -
+traffic to any port on the node, will be forwarded to its endpoints on the same port. This could potentially break networking on the node, if traffic for, say, port 22 got forwarded to \<endpoint IP\>:22.
+
+In order to expose the entire port range on a supported service, a user needs to:
+1) Create a single ServicePort entry in the ServiceSpec and set "allPorts" to True.
+2) Populate the protocol field.
+3) Populate a valid port in the port field.
+   This is required to make sure we do not break other Service API clients (like DNS) that might always expect a non-zero port value. The [APIServer validation](https://github.com/kubernetes/kubernetes/blob/b9ce4ac212d150212485fa29d62a2fbd783a57b0/pkg/apis/core/validation/validation.go#L4373) could be relaxed, but we might break other clients.
+
+By default, a NodePort will assigned for the single valid port in the service, if it is of Service Type=LoadBalancer. Traffic Forwarding will work as usual for that NodePort.
+NodePort allocation can be disabled by setting "allocateLoadBalancerNodePorts: false" on the service spec.
+
+### Risks and Mitigations
+
+A user could accidentally expose the entire port range on their cluster nodes, by enabling it for a Service. This is mostly applicable to LoadBalancer services, that are typically
+accessible from outside the cluster.
+The behavior depends on the firewall implementation in the environment. Users should make sure that traffic to NodeIP:\<servicePor\t> is disallowed by the right firewall rules.  
+If such firewalling does not exist, it is possible that ports on the nodes are exposed via Services even today, but will allPorts, it becomes trivial to accidentally expose
+the entire port range.
+
+Currently, kube-proxy adds the right rules to only allow traffic to the ServiceIP/Port combination specified in the service. When using allPorts, kube-proxy will allow all traffic for the given protocol and given serviceIP. Kube-Proxy alone cannot drop traffic to NodeIP:\<servicePort\>.
+
+## Design Details
+
+Changes are required to APIServer validation, kube-proxy and controllers that use the ServicePort field.
+
+1) New Validation checks:
+
+   * allPorts can be set to true only for ClusterIP and LoadBalancer services.
+   * Only one ServicePort object per protocol can be specified in the ServiceSpec, if allPorts is set to true.
+
```

Comment:
Can the service mutate from ClusterIP to LoadBalancer?
Is `allPorts` immutable once set?

Reply:
Yes, the service can mutate, allPorts field will be mutable. Since it is not specified in the validation, the default behavior is that mutation is allowed.

Reply:
Captured this in the spec, resolving.